contracts/Staking.sol

```diff
@@ -0,0 +1,559 @@
+    _;
+  }
+
+  modifier initParams() {
+    if (!alreadyInit) {
+      oracleRelayerFee = INIT_ORACLE_RELAYER_FEE;
+      minDelegation = INIT_MIN_DELEGATION;
+      alreadyInit = true;
+    }
+    _;
+  }
+
+  /*********************************** Events **********************************/
+  event delegateSubmitted(address indexed delegator, address indexed validator, uint256 amount, uint256 oracleRelayerFee);
+  event undelegateSubmitted(address indexed delegator, address indexed validator, uint256 amount, uint256 oracleRelayerFee);
+  event redelegateSubmitted(address indexed delegator, address indexed validatorSrc, address indexed validatorDst, uint256 amount, uint256 oracleRelayerFee);
+  event rewardReceived(address indexed delegator, uint256 amount);
+  event rewardClaimed(address indexed delegator, uint256 amount);
+  event undelegatedReceived(address indexed delegator, address indexed validator, uint256 amount);
+  event undelegatedClaimed(address indexed delegator, uint256 amount);
+  event failedDelegate(address indexed delegator, address indexed validator, uint256 amount, uint8 errCode);
+  event failedUndelegate(address indexed delegator, address indexed validator, uint256 amount, uint8 errCode);
+  event failedRedelegate(address indexed delegator, address indexed valSrc, address indexed valDst, uint256 amount, uint8 errCode);
+  event paramChange(string key, bytes value);
+  event failedSynPackage(uint8 indexed eventCode, uint256 errCode);
+  event crashResponse(uint8 indexed eventCode);
+
+  receive() external payable {}
+
+  /************************* Implement cross chain app *************************/
+  function handleSynPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract initParams override returns(bytes memory) {
+    RLPDecode.Iterator memory iter = msgBytes.toRLPItem().iterator();
+    uint8 eventCode = uint8(iter.next().toUint());
+    uint32 resCode;
+    if (eventCode == EVENT_DISTRIBUTE_REWARD) {
+      resCode = _handleDistributeRewardSynPackage(iter);
+    } else if (eventCode == EVENT_DISTRIBUTE_UNDELEGATED) {
+      resCode = _handleDistributeUndelegatedSynPackage(iter);
+    } else {
+      resCode = ERROR_UNKNOWN_PACKAGE_TYPE;
+    }
+
+    if (resCode == CODE_OK) {
+      return new bytes(0);
+    } else {
+      emit failedSynPackage(eventCode, resCode);
+      return _RLPEncode(eventCode, msgBytes);
+    }
+  }
+
+  function handleAckPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract initParams override {
+    RLPDecode.Iterator memory iter = msgBytes.toRLPItem().iterator();
+    uint8 eventCode = uint8(iter.next().toUint());
+    if (eventCode == EVENT_DELEGATE) {
+      _handleDelegateAckPackage(iter);
+    } else if (eventCode == EVENT_UNDELEGATE) {
+      _handleUndelegateAckPackage(iter);
+    } else if (eventCode == EVENT_REDELEGATE) {
+      _handleRedelegateAckPackage(iter);
+    } else {
+      require(false, "unknown event type");
```

Comment:
Should not panic here, because this is a valid cross chain package(because it pass the merkel proof verification), and this is the last handler(can not answer a ack package), we just emit event here, no panic.

Reply:
Fixed